### PR TITLE
[ROCm] Create torchvision as a HIP Extension

### PR DESCRIPTION
**Pull request by ashishfarmer**
This pull request enables creation of torchvision as a pytorch extension on ROCm builds. Major changes in this PR are:
1. Adding a hipification step in setup.py before building of torchvision when running on ROCm
2. Adding a way to build the extension on ROCm
3. Adding minor modifications in the source code to facilitate hipification

cc: @ezyang @iotamudelta

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -83,7 +84,24 @@ def get_extensions():
 
     main_file = glob.glob(os.path.join(extensions_dir, '*.cpp'))
     source_cpu = glob.glob(os.path.join(extensions_dir, 'cpu', '*.cpp'))
-    source_cuda = glob.glob(os.path.join(extensions_dir, 'cuda', '*.cu'))
+
+    is_rocm_pytorch = True if ROCM_HOME is not None else False
```
Comment (iotamudelta) [inline setup.py]: Can we also here use the hip version @t-vi added?

Comment (ashishfarmer) [reply setup.py]: Good catch. Added it, now it is a more reliable check

File: setup.py [hunk 2]
```diff
@@ -12,7 +12,8 @@
 import shutil
 
 import torch
-from torch.utils.cpp_extension import BuildExtension, CppExtension, CUDAExtension, CUDA_HOME
+from torch.utils.cpp_extension import BuildExtension, CppExtension, CUDAExtension, CUDA_HOME, ROCM_HOME
```
Comment (fmassa) [inline setup.py]: Looks like you'll need to guard the `ROCM_HOME` import somehow, either in torchvision or in PyTorch (I think it would be preferable to do it in PyTorch)

Comment (t-vi) [reply setup.py]: Indeed. The easiest probably would be to set `ROCM_HOME` to `None` if it isn't available, just as we do with `CUDA_HOME`. This is in `torch.utils.cpp_extension` itself.

Comment (ashishfarmer) [reply setup.py]: I think `ROCM_HOME` value is already set to `None` in pyTorch when it is not built for ROCm. For the CI tests that are running into the error `ImportError: cannot import name 'ROCM_HOME' from 'torch.utils.cpp_extension' `, I noticed that the tests install `torch-1.4.0`. The support of extensions on ROCm went into pyTorch very recently, so it looks like it is trying to read `ROCM_HOME` from pyTorch version it was not put in. 
The CI tests here that are passing, all install `torch-1.5.0.+` and can see the variable.

Comment (ashishfarmer) [reply setup.py]: @fmassa @t-vi would you suggest checking for pyTorch version in torchvision for ROCm build?

Comment (t-vi) [reply setup.py]: I know the issue came up in previous PRs, but I'm not sure what the conclusion was...

Comment (ashishfarmer) [reply setup.py]: I just added a check for torch version around reading `ROCM_HOME`. That should take care of a scenario with older version of pytorch trying to use torchvision with ROCm enabled

Comment (fmassa) [reply setup.py]: > I noticed that the tests install torch-1.4.0

This is not quite true I believe, we use PyTorch nightly in our CI, see https://github.com/pytorch/vision/blob/7d1cd1de34298ccb4a993d7c6af59b08cf8ac094/.travis.yml#L37 and https://github.com/pytorch/vision/blob/7d1cd1de34298ccb4a993d7c6af59b08cf8ac094/.circleci/config.yml.in#L42
(or, for simplicity, see the output of one of the CI jobs, like https://circleci.com/gh/pytorch/vision/96101?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link, which shows
```
pytorch:         1.5.0.dev20200305-py3.5_cuda10.0.130_cudnn7.6.3_0 pytorch-nightly
```
)

The change looks ok to me, but I think it can be simplified as we assume that torchvision nightly should use a PyTorch nightly.

Here is what I propose, we can merge this as is and then can you send a follow-up PR cleaning up the checking logic for PyTorch?

**Comment by seemethere** [comment]
So the current situation with ROCm CI is a tricky one.

Basically we want our single CI solution to be CircleCI for the time being but they do not currently have machines that can run ROCm or have the ability to connect our own machines that can run ROCm.

We're trying extremely hard to get rid of Jenkins and I think adding this project to Jenkins would just dig ourselves a deeper hole with it.

**TL;DR** 

>I would love to add ROCm support for our CI through CircleCI, but until we can do that I don't feel like the effort to bring this to Jenkins would be worth it.

**Comment by iotamudelta** [comment]
@seemethere that seems reasonable. I also think once we merge this in and push the necessary cmake change for torchvision GPU kernels on ROCm to PyTorch and its CI, we will at least have some point testing there?

**Comment by fmassa** [comment]
@iotamudelta 

>  I also think once we merge this in and push the necessary cmake change for torchvision GPU kernels on ROCm to PyTorch and its CI, we will at least have some point testing there?

I'm not sure I understand, but in the cases where we add torchvision tests running in PyTorch, we keep the torchvision version fixed and only updated it occasionally. This wouldn't help avoiding breaking changes at commit time, although it would give some (occasional) signal.

I'm merging this as is, but as I said without CI it will be hard to guarantee that things work ok on ROCm.

**Comment by iotamudelta** [comment]
@fmassa thanks! I was referring to that testing of certain integration points. We can revisit proper CI once the circleci situation is clearer on our side.
